File: uniconn-go/webtransport/conn.go
package webtransport

import (
	"net"
	"time"

	wt "github.com/quic-go/webtransport-go"
)

// addr implements net.Addr for WebTransport connections.
type addr struct {
	network string
	str     string
}

func (a *addr) Network() string { return a.network }
func (a *addr) String() string  { return a.str }

// Conn wraps a *webtransport.Session + *webtransport.Stream to implement net.Conn.
// Following the "one connection = one session = one bidirectional stream" model.
type Conn struct {
	session *wt.Session
	stream  *wt.Stream
}

// WrapConn creates a net.Conn from a WebTransport session and stream.
func WrapConn(session *wt.Session, stream *wt.Stream) net.Conn {
	return &Conn{session: session, stream: stream}
}

func (c *Conn) Read(b []byte) (int, error) {
	return c.stream.Read(b)
}

func (c *Conn) Write(b []byte) (int, error) {
	return c.stream.Write(b)
}

// Close closes the stream and the underlying WebTransport session.
func (c *Conn) Close() error {
	c.stream.CancelRead(0)
	c.stream.CancelWrite(0)
	return c.session.CloseWithError(0, "closed")
}

func (c *Conn) LocalAddr() net.Addr {
	return &addr{
		network: "webtransport",
		str:     c.session.LocalAddr().String(),
	}
}

func (c *Conn) RemoteAddr() net.Addr {
	return &addr{
		network: "webtransport",
		str:     c.session.RemoteAddr().String(),
	}
}

func (c *Conn) SetDeadline(t time.Time) error {
	return c.stream.SetDeadline(t)
}

func (c *Conn) SetReadDeadline(t time.Time) error {
	return c.stream.SetReadDeadline(t)
}

func (c *Conn) SetWriteDeadline(t time.Time) error {
	return c.stream.SetWriteDeadline(t)
}

File: uniconn-js/web/src/webtransport/conn.ts
/**
 * WebTransport browser client adapter.
 *
 * This module provides an IConn implementation using the browser's
 * native WebTransport API. It wraps a single bidirectional stream
 * from a WebTransport session into the IConn interface.
 *
 * NOTE: This module only works in browser environments that support
 * the WebTransport API (Chrome 97+, Edge 97+).
 * Server-side WebTransport for Node.js is marked as TODO.
 */

import type { Addr, IConn, IDialer, DialOptions } from "@uniconn/core";
import { ConnectionClosedError, TimeoutError } from "@uniconn/core";

/**
 * WebTransportConn wraps a WebTransport bidirectional stream into IConn.
 */
export class WebTransportConn implements IConn {
  private session: WebTransport;
  private reader: ReadableStreamDefaultReader<Uint8Array>;
  private writer: WritableStreamDefaultWriter<Uint8Array>;
  private closed = false;
  private readDeadlineMs = 0;
  private writeDeadlineMs = 0;
  private pendingData: Uint8Array | null = null;
  private pendingOffset = 0;
  private url: string;

  constructor(
    session: WebTransport,
    reader: ReadableStreamDefaultReader<Uint8Array>,
    writer: WritableStreamDefaultWriter<Uint8Array>,
    url: string,
  ) {
    this.session = session;
    this.reader = reader;
    this.writer = writer;
    this.url = url;
  }

  async read(buffer: Uint8Array): Promise<number> {
    if (this.closed) throw new ConnectionClosedError();

    // Consume leftover from previous read
    if (this.pendingData) {
      const remaining = this.pendingData.length - this.pendingOffset;
      const n = Math.min(remaining, buffer.length);
      buffer.set(
        this.pendingData.subarray(this.pendingOffset, this.pendingOffset + n),
      );
      this.pendingOffset += n;
      if (this.pendingOffset >= this.pendingData.length) {
        this.pendingData = null;
        this.pendingOffset = 0;
      }
      return n;
    }

    const readPromise = this.reader.read().then(({ done, value }) => {
      if (done || !value) return 0;
      const n = Math.min(value.length, buffer.length);
      buffer.set(value.subarray(0, n));
      if (n < value.length) {
        this.pendingData = value;
        this.pendingOffset = n;
      }
      return n;
    });

    if (this.readDeadlineMs > 0) {
      return Promise.race([
        readPromise,
        new Promise<number>((_, reject) =>
          setTimeout(() => reject(new TimeoutError("read")), this.readDeadlineMs),
        ),
      ]);
    }

    return readPromise;
  }

  async write(data: Uint8Array): Promise<number> {
    if (this.closed) throw new ConnectionClosedError();

    const writePromise = this.writer.write(data).then(() => data.length);

    if (this.writeDeadlineMs > 0) {
      return Promise.race([
        writePromise,
        new Promise<number>((_, reject) =>
          setTimeout(
            () => reject(new TimeoutError("write")),
            this.writeDeadlineMs,
          ),
        ),
      ]);
    }

    return writePromise;
  }

  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    try {
      this.reader.releaseLock();
    } catch { /* ignore */ }
    try {
      await this.writer.close();
    } catch { /* ignore */ }
    this.session.close();
  }

  localAddr(): Addr {
    return { network: "webtransport", address: "browser" };
  }

  remoteAddr(): Addr {
    return { network: "webtransport", address: this.url };
  }

  setDeadline(ms: number): void {
    this.readDeadlineMs = ms;
    this.writeDeadlineMs = ms;
  }

  setReadDeadline(ms: number): void {
    this.readDeadlineMs = ms;
  }

  setWriteDeadline(ms: number): void {
    this.writeDeadlineMs = ms;
  }
}

/**
 * WebTransportDialer implements IDialer for browser WebTransport connections.
 *
 * This creates a WebTransport session and opens a single bidirectional stream,
 * following the "one connection = one session = one stream" model.
 */
export class WebTransportDialer implements IDialer {
  async dial(address: string, options?: DialOptions): Promise<IConn> {
    const transport = new WebTransport(address);

    if (options?.timeout) {
      const timer = setTimeout(() => transport.close(), options.timeout);
      transport.ready.then(() => clearTimeout(timer));
    }

    if (options?.signal) {
      if (options.signal.aborted) {
        transport.close();
        throw new Error("dial aborted");
      }
      options.signal.addEventListener(
        "abort",
        () => transport.close(),
        { once: true },
      );
    }

    await transport.ready;

    const stream = await transport.createBidirectionalStream();
    const reader = stream.readable.getReader();
    const writer = stream.writable.getWriter();

    return new WebTransportConn(transport, reader, writer, address);
  }
}
